### ISSUE TITLE
Gradient of scipy.sparse.spsolve

### ISSUE THREAD

**Issue by twhughes** [HIPS/autograd#433]
I'm interested in using autograd for applications where one has an objective `J(x)` where the vector `x` is the solution to the sparse linear system `Ax = b`.

How could one implement the gradient of `x = spsolve(A, b)` using autograd?

It is well known that `dx/dp = d/dp [ inv(A) b ] = -inv(A) dA/dp x` but I am not sure how to translate this into autograd function.

Are there any alternatives?  Perhaps if the matrix A is in LU decomposition form one may use `solve_triangular`  from `scipy.lilnalg`?

Thanks

Tyler

**Comment by plisdku** [comment]
Sounds like photonic inverse design?

I tried this (not for photonic inverse design) a couple weeks ago.  I wrapped up the sparse matrix operations in primitives so autograd didn't have to see them, and sent in the val+row+col data as the arguments to the linear system solver.  This way I was able to get only the gradient with respect to the nonzeros in the matrix, which is of course what you would want (prevent explosion into a dense vector).

I think it might work very nicely for permittivities.  In my experience runtime will depend on how efficiently you can create the nonzero values.  In my application I couldn't figure out a way to make the values efficiently and the runtime was horrific.

**Comment by twhughes** [comment]
@plisdku Yes this is exactly right :D  If you have that code I'd love to see it.  The plan was to write some package for doing fully automated adjoint-based optimization of photonic structures.  Perhaps we could chat some time about it.

**Comment by twhughes** [comment]
see also https://github.com/HIPS/autograd/issues/80#issuecomment-429158923

**Comment by shoyer** [comment]
> I wrapped up the sparse matrix operations in primitives so autograd didn't have to see them, and sent in the val+row+col data as the arguments to the linear system solver. This way I was able to get only the gradient with respect to the nonzeros in the matrix, which is of course what you would want (prevent explosion into a dense vector).

Yes, this works very nicely. For the record (I already showed this to @twhughes) you can find a version of this as an Autograd primitive here:
https://github.com/google-research/neural-structural-optimization/blob/07ce57f9045ed7b32fd44e67c546352dc9034f60/neural_structural_optimization/autograd_lib.py#L156-L203
